### DISCUSSION TITLE
[labs/context] Can providers use the value they provide?

### DISCUSSION THREAD

**Discussion by joezappie** [lit/lit#3804] [Q&A] [answered]
I have an element that is providing some async loaded data from an API. I have children inside the element consuming the context, which is working.

I need my provider to also use the data being provided in the context, but it the property just comes back undefined?

```
import { userContext } from './contexts/user.js';

export class UserView extends LitElement {
  @property({ attribute: false })
  @provide({ context: userContext })
  user;

  render() {
    return html`
      <user-contact></user-contact>
      <div>${this.user?.name}</div>
    `;
  }

  async connectedCallback() {
    super.connectedCallback();
    const user = await App.client.service('users').findOne(this.id});
    this.user = user;
    console.log('provider', 'user', user) // not undefined
    console.log('provider', 'user', this.user) // is undefined?
  }
}
```

Logging out the loaded user, its `undefined`. The child user-contact that `@consumes` the `userContext`, gets the data.

Is this correct? The documentation show adding `@property` along with `@provide`, so why isn't the value available in the provider element via `this.user`?

**Comment by augustjk** [comment]
You need the flip the order of the decorators.

```ts
export class UserView extends LitElement {
  @provide({ context: userContext })
  @property({ attribute: false })
  user;

  ...
}
```

**Reply by joezappie** [reply]
Hmm, so I did try that originally and while I can then use `this.user` in the provider's class, it does seem to break consuming it in children elements:

```
import { userContext } from './contexts/user.js';

export class UserContact extends LitElement {
  @property({ attribute: false })
  @consume({ context: userContext, subscribe: true })
  user;

  render() {
    console.log('consumer', 'user', this.user); // logs undefined
    return nothing;
    `;
  }
```

If I flip `@property` to be after `@consume` in my UserContact element, I get an error `consume.js:50 Uncaught TypeError: Cannot assign to read only property 'Symbol()' of object '#<UserContact>'` so I don't think that's right.

I do notice that the `render` for UserContact is only called once after flipping the decorators in the provider class. I get the following:

```
consumer user undefined
provider user { name: ... }
```

So it doesn't seem like its getting the update that the value has changed even though I subscribed to it. Child works fine when the decorators are in the order originally posted.

**Reply by joezappie** [reply]
Even weirder, everything works in this sandbox I made: https://codesandbox.io/s/nameless-violet-ukw7yp?file=/src/my-component.ts:0-863

Copying that same code and putting it into my project, the child does not get updated with the phone number. I've double checked my versions are the same as in the sandbox:

```
window.litElementVersions
['3.3.1']
window.litHtmlVersions
['2.7.2']
```

Only real difference I can think of is that sandbox is in typescript while my project is in javascript so not sure if theres a difference in how the decorators work in either environment.

**Reply by joezappie** [reply]
I setup a codesandbox.io with normal JS and babel's decorator proposal (same config as my local environment) and the issue is reproducible. Seems like there is a difference between how the TS and JS decorators work.

https://codesandbox.io/s/friendly-austin-lbrx4v?file=/src/index.js

**Reply by augustjk** [reply]
Hmm you're totally right. It looks like specifically the `@provide` decorator does not work with babel, but replacing it with the `ContextProvider` does work.
https://codesandbox.io/s/old-architecture-vvkx0x

**Reply by augustjk** [reply]
It is actually really weird. `@provide` decorator does not work on it's own, but if you combine with the `@property` decorator before it like your original, it does seem to provide the context.

**Reply by joezappie** [reply]
Glad I'm not just crazy. I was so confused on why it didn't work with the examples in the readme but did (kinda) work with property declared first. 

Would you like me to open an issue for this? I think I read the decorators are kind of on a development freeze until the decorator proposal is finalized, but does this warrant a fix? My workaround is just to have my providing class have two properties so its not a roadblock but would be nice to not have to do this:

```
@provide({ context: userContext })
@property({ attribute: false })
userProvider;

@property({ attribute: false }) 
user;

async connectedCallback() {
    super.connectedCallback();
    this.user = await App.client.service('users').findOne(this.id});
    this.userProvider = this.user;
  }
``

**Reply by augustjk** [reply, marked as answer]
I filed it here https://github.com/lit/lit/issues/3806
Thanks for reporting and narrowing down the repro!

I'll try and take a look later to see if it might be a quick fix but since there is a workaround priority is not high.

Another workaround would be to use the `ContextProvider` reactive controller as well. The double decorator seems less desirable as a workaround as it feels like relying on buggy behavior to make the decorator work.
